File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        res = 0
        stack = [-101]

        def traverse(root):
            nonlocal res
            top_val = stack[-1]

            if not root:
                return

            if root.val >= top_val:
                stack.append(root.val)
                res += 1

            left = traverse(root.left)
            right = traverse(root.right)

            if root.val >= top_val:
                stack.pop(-1)

        traverse(root)
        return res


# do DFS
# before calling next root, save value to stack
# check if value is smaller than biggest, and if bigger then add 1 to res
# 

File: Data Structures & Algorithms/invert-a-binary-tree/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// if 
// if left not null, go left
// if right not null, go right
// return and reverse children

// if left smaller than root go left
// if left null && right not null go right
// reverse children



// if root null, return null
// if left smaller than right, revert node children
// return smaller node 

// class Solution {
//     public TreeNode invertTree(TreeNode root) {
//         if (root == null) {
//             return;
//         }
//         if (root.left.val < root.val){ // reverse children
//             int oldLeft = root.left.val;
//             root.left.val = root.right.val;
//             root.right.val = oldLeft;
//             return invertTree(root.left);
//         }
//         if (root.right.val )
//         return root;
//     }
// }

class Solution {
    public TreeNode invertTree(TreeNode root) {
        if (root == null) {
            return null;
        }
        TreeNode left = invertTree(root.left);
        TreeNode right = invertTree(root.right);

        root.left = right;
        root.right = left;

        return root;
    }
}